plugins/ccgx/fu-ccgx-hpi-device.c

```diff
@@ -605,11 +1068,60 @@ fu_ccgx_hpi_write_firmware (FuDevice *device,
 			    FwupdInstallFlags flags,
 			    GError **error)
 {
-	g_set_error_literal (error,
-			     FWUPD_ERROR,
-			     FWUPD_ERROR_NOT_SUPPORTED,
-			     "not supported");
-	return FALSE;
+	FuCcgxHpiDevice *self = FU_CCGX_HPI_DEVICE (device);
+	CCGxMetaData metadata = { 0x0 };
+	GPtrArray *records = fu_ccgx_firmware_get_records (FU_CCGX_FIRMWARE (firmware));
+	FWMode fw_mode_alt = fu_ccgx_fw_mode_get_alternate (self->fw_mode);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* enter flash mode */
+	locker = fu_device_locker_new_full (self,
+					    (FuDeviceLockerFunc) fu_ccgx_hpi_enter_flash_mode,
+					    (FuDeviceLockerFunc) fu_ccgx_hpi_leave_flash_mode,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* invalidate metadata for alternate image */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+	if (!fu_ccgx_hpi_load_metadata (self, fw_mode_alt, &metadata, error))
+		return FALSE;
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_ERASE);
+	metadata.metadata_valid = 0x00;
+	if (!fu_ccgx_hpi_save_metadata (self, fw_mode_alt, &metadata, error))
+		return FALSE;
+
+	/* write new image */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	for (guint i = 0; i < records->len; i++) {
+		FuCcgxFirmwareRecord *rcd = g_ptr_array_index (records, i);
+
+		/* write chunk */
+		if (!fu_ccgx_hpi_write_flash (self, rcd->row_number,
+					      g_bytes_get_data (rcd->data, NULL),
+					      g_bytes_get_size (rcd->data),
+					      error)) {
+			g_prefix_error (error, "fw write error @0x%x: ", rcd->row_number);
+			return FALSE;
+		}
+
+		/* update progress */
+		fu_device_set_progress_full (device, (gsize) i, (gsize) records->len);
+	}
+
+	/* validate fw */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_VERIFY);
+	if (!fu_ccgx_hpi_validate_fw (self, fw_mode_alt, error)) {
+		g_prefix_error (error, "fw validate error: ");
+		return FALSE;
+	}
+
+	/* this is a good time to leave the flash mode *before* rebooting */
+	g_clear_object (&locker);
```

Comment:
why not directly call `fu_ccgx_hpi_leave_flash_mode` so you can do error handling in case of a failure here?

Reply:
The idea of doing it as a `FuDeviceLocker` was that we would disable the flash mode on all the error paths too. We could certainly do as you suggest, but we'd need to add something like `gboolean fu_device_locker_close(FuDeviceLocker *self, GError **error)` to not _close_ things twice. I'm happy to do that if you think it's a good idea. Better names welcome too.

Reply:
Yes, I think that an explicit close makes more sense.  That name seems fine to me.

Reply:
https://github.com/fwupd/fwupd/pull/1916